File: code/class-05-polymorphism/polymorphism_demo.py
"""
polymorphism_demo.py

Class 05 – Polymorphism

Demonstrates:
  - Runtime polymorphism via method overriding
  - Duck typing (Python's natural form of polymorphism)
  - isinstance() checks
  - A list of mixed objects treated uniformly
"""
import math


# ── Base class ────────────────────────────────────────────────────────────────

class Shape:
    """Abstract-like base for geometric shapes."""

    def __init__(self, color: str) -> None:
        self.color = color

    def area(self) -> float:
        raise NotImplementedError("Subclasses must implement area()")

    def perimeter(self) -> float:
        raise NotImplementedError("Subclasses must implement perimeter()")

    def describe(self) -> None:
        print(
            f"{type(self).__name__:<12} [color={self.color}]"
            f"  area={self.area():.2f}  perimeter={self.perimeter():.2f}"
        )


# ── Concrete shapes ───────────────────────────────────────────────────────────

class Circle(Shape):
    def __init__(self, color: str, radius: float) -> None:
        super().__init__(color)
        self.radius = radius

    def area(self)      -> float: return math.pi * self.radius ** 2
    def perimeter(self) -> float: return 2 * math.pi * self.radius


class Rectangle(Shape):
    def __init__(self, color: str, width: float, height: float) -> None:
        super().__init__(color)
        self.width  = width
        self.height = height

    def area(self)      -> float: return self.width * self.height
    def perimeter(self) -> float: return 2 * (self.width + self.height)


class Triangle(Shape):
    def __init__(self, color: str, a: float, b: float, c: float) -> None:
        super().__init__(color)
        self.a, self.b, self.c = a, b, c

    def area(self) -> float:
        s = self.perimeter() / 2
        return math.sqrt(s * (s - self.a) * (s - self.b) * (s - self.c))  # Heron

    def perimeter(self) -> float: return self.a + self.b + self.c


# ── Duck typing demo ──────────────────────────────────────────────────────────

class Printer:
    """Overloaded print – Python handles this via default args / *args."""

    def print_value(self, value) -> None:
        print(f"{type(value).__name__}: {value}")


# ── Demo ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    # Polymorphic collection
    shapes: list[Shape] = [
        Circle("red",    5.0),
        Rectangle("blue",  4.0, 6.0),
        Triangle("green",  3.0, 4.0, 5.0),
    ]

    print("=== All Shapes ===")
    for shape in shapes:
        shape.describe()    # correct method called at runtime

    # isinstance check
    print("\n=== isinstance ===")
    for shape in shapes:
        print(f"{type(shape).__name__} is Shape: {isinstance(shape, Shape)}")

    # Duck typing – anything with area() and perimeter() works
    print("\n=== Duck Typing Demo ===")

    class HexagonDuck:
        """Not a subclass of Shape, but has the same interface."""
        def area(self)      -> float: return 37.1
        def perimeter(self) -> float: return 24.0

    duck = HexagonDuck()
    for obj in [Circle("purple", 2.0), duck]:
        print(f"area={obj.area():.2f}  perimeter={obj.perimeter():.2f}")

    # Overloading via default args
    print("\n=== Printer ===")
    p = Printer()
    p.print_value(42)
    p.print_value(3.14)
    p.print_value("Hello OOP")

File: code/class-05-polymorphism/PolymorphismDemo.java
/**
 * PolymorphismDemo.java
 *
 * Class 05 – Polymorphism
 *
 * Demonstrates:
 *   - Runtime polymorphism (method overriding + dynamic dispatch)
 *   - Compile-time polymorphism (method overloading)
 *   - Upcasting and downcasting
 *   - Polymorphic collection iteration
 */
import java.util.ArrayList;
import java.util.List;

// ===================== Base class =====================
abstract class Shape {
    protected String color;

    public Shape(String color) { this.color = color; }

    // Each subclass MUST implement this (runtime polymorphism)
    public abstract double area();
    public abstract double perimeter();

    public void describe() {
        System.out.printf("%s [color=%s, area=%.2f, perimeter=%.2f]%n",
                getClass().getSimpleName(), color, area(), perimeter());
    }
}

// ===================== Circle =====================
class Circle extends Shape {
    private double radius;

    public Circle(String color, double radius) {
        super(color);
        this.radius = radius;
    }

    @Override public double area()      { return Math.PI * radius * radius; }
    @Override public double perimeter() { return 2 * Math.PI * radius; }
}

// ===================== Rectangle =====================
class Rectangle extends Shape {
    private double width;
    private double height;

    public Rectangle(String color, double width, double height) {
        super(color);
        this.width  = width;
        this.height = height;
    }

    @Override public double area()      { return width * height; }
    @Override public double perimeter() { return 2 * (width + height); }
}

// ===================== Triangle =====================
class Triangle extends Shape {
    private double a, b, c;   // three sides

    public Triangle(String color, double a, double b, double c) {
        super(color);
        this.a = a; this.b = b; this.c = c;
    }

    @Override public double area() {
        double s = perimeter() / 2;
        return Math.sqrt(s * (s - a) * (s - b) * (s - c)); // Heron's formula
    }

    @Override public double perimeter() { return a + b + c; }
}

// ===================== Overloading demo =====================
class Printer {
    public void print(int value)    { System.out.println("int: "    + value); }
    public void print(double value) { System.out.println("double: " + value); }
    public void print(String value) { System.out.println("String: " + value); }
}

// ===================== Demo =====================
public class PolymorphismDemo {

    public static void main(String[] args) {

        // --- Polymorphic collection ---
        List<Shape> shapes = new ArrayList<>();
        shapes.add(new Circle("red",    5.0));
        shapes.add(new Rectangle("blue", 4.0, 6.0));
        shapes.add(new Triangle("green", 3.0, 4.0, 5.0));

        System.out.println("=== All Shapes ===");
        for (Shape s : shapes) {
            s.describe();   // correct implementation called at runtime
        }

        // --- Upcasting / Downcasting ---
        System.out.println("\n=== Casting ===");
        Shape s = new Circle("yellow", 3.0); // upcasting
        if (s instanceof Circle c) {         // pattern matching instanceof (Java 16+)
            System.out.println("Radius available after downcast: use Circle API here");
            c.describe();
        }

        // --- Compile-time overloading ---
        System.out.println("\n=== Overloading ===");
        Printer p = new Printer();
        p.print(42);
        p.print(3.14);
        p.print("Hello OOP");
    }
}
